src/main/java/pl/edu/pw/mini/ingreedio/api/service/ProductService.java

```diff
@@ -40,8 +70,94 @@ public Product addProduct(Product product) {
         return productRepository.save(product);
     }
 
-    public Page<Product> getProductsMatching(ProductFilterCriteria criteria,
-                                                Pageable pageable) {
-        return productRepository.getProductsMatching(criteria, pageable);
+    public Page<ProductDto> getProductsMatching(ProductFilterCriteria criteria, Pageable pageable) {
+        Page<Product> productsPage = productRepository.getProductsMatching(criteria, pageable);
+
+        String u = authService.getCurrentUsername();
+
+        Optional<User> userOptional = userService
+            .getUserByUsername(authService.getCurrentUsername());
+
+        if (userOptional.isPresent()) {
+            User user = userOptional.get();
+            Long userId = user.getId();
+
+            List<ProductDto> productDtos = productsPage.getContent().stream()
+                .map(product -> {
+                    Boolean isLiked = product.getLikedBy() != null
+                        ? product.getLikedBy().contains(userId) : null;
+                    return ProductDto.builder()
+                        .id(product.getId())
+                        .name(product.getName())
+                        .smallImageUrl(product.getSmallImageUrl())
+                        .provider(product.getProvider())
+                        .shortDescription(product.getShortDescription())
+                        .isLiked(isLiked)
+                        .build();
+                })
+                .collect(Collectors.toList());
+
+            return new PageImpl<>(productDtos,
+                productsPage.getPageable(),
+                productsPage.getTotalElements());
+        }
+
+        return productsPage.map(productDtoMapper);
+    }
+
+    public boolean likeProduct(Long productId) {
+        Optional<User> userOptional = userService
+            .getUserByUsername(authService.getCurrentUsername());
+        if (userOptional.isPresent()) {
+            User user = userOptional.get();
+            Long userId = user.getId();
+
+            Optional<Product> productOptional = productRepository.findById(productId);
+            if (productOptional.isPresent()) {
+                Product product = productOptional.get();
+                List<Long> likedBy = product.getLikedBy();
+
+                if (likedBy == null) {
+                    likedBy = new ArrayList<>();
+                }
+
+                if (!likedBy.contains(userId)) {
+                    likedBy.add(userId);
+                    product.setLikedBy(likedBy);
+                    productRepository.save(product);
+                    userService.likeProduct(userId.intValue(), productId);
+                }
+                return true;
+            }
+        }
+        return false;
```

Comment:
We can improve readability of this code by reducing nesting:
```suggestion
        Optional<User> userOptional = userService
            .getUserByUsername(authService.getCurrentUsername());
        if (userOptional.isEmpty()) {
            return false;
        }

        Long userId = userOptional.get().getId();

        Optional<Product> productOptional = productRepository.findById(productId);
        if (productOptional.isEmpty()) {
            return false;
        }

        Product product = productOptional.get();
        List<Long> likedBy = product.getLikedBy();

        if (likedBy == null) {
            likedBy = new ArrayList<>();
        }

        if (!likedBy.contains(userId)) {
            likedBy.add(userId);
            product.setLikedBy(likedBy);
            productRepository.save(product);
            userService.likeProduct(userId.intValue(), productId);
        }
        return true;
```